File: internal/turnstile/turnstile.go
package turnstile

import (
	"github.com/gofiber/fiber/v3"
	"github.com/gofiber/fiber/v3/client"
	"github.com/mayocream/twitter/internal/config"
)

const TURNSTILE_URL = "https://challenges.cloudflare.com/turnstile/v0/siteverify"

type Turnstile struct {
	Config *config.Config
}

type TurnstileResponse struct {
	Success bool `json:"success"`
}

func NewTurnstile(config *config.Config) *Turnstile {
	return &Turnstile{
		Config: config,
	}
}

func (t *Turnstile) Verify(token string) error {
	result, err := client.Post(TURNSTILE_URL, client.Config{
		FormData: map[string]string{
			"secret":   t.Config.TurnstileSecretKey,
			"response": token,
		},
	})
	if err != nil {
		return err
	}

	if result.StatusCode() != fiber.StatusOK {
		return fiber.ErrForbidden
	}

	var response TurnstileResponse
	if err := result.JSON(&response); err != nil {
		return err
	}

	if !response.Success {
		return fiber.ErrForbidden
	}

	return nil
}

File: lib/turnstile.ts
export const verify = async (token: string) => {
  const response = await fetch(
    'https://challenges.cloudflare.com/turnstile/v0/siteverify',
    {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        response: token,
        secret: process.env.TURNSTILE_SECRET_KEY!,
      }),
    },
  )

  if (!response.ok) {
    console.error('Failed to contact Turnstile')
    return false
  }

  const data = await response.json()
  return !!data.success
}
